### DISCUSSION TITLE
looking for an example for running codes on ddr of 4 harts (bare metal)  "mpfs-hal-run-from-ddr-u54-1-u54-2 -u54-3 -u54_4"

### DISCUSSION THREAD

**Discussion by kevinwen2000** [polarfire-soc/polarfire-soc-documentation#256] [Applications and demos] [answered]
I found a design example called mpfs-hal-run-from-ddr-u54-1-u54-2. Two programs (u54_1.c and u54_2.c) are running on u_54_1 and u54_2 from ddr. It is successful for running on ddr memory after generating payload with yaml configuration file. 
The question is I don’t know what the correct steps is for building 4 different bare metal programs running from ddr of u54_1 u54_2, u54_3 and u54_4 respectively. 
Should we create 4 elf files from 4 independent projects to generate payload.bin (yaml example?) and how to set the MPFS_HAL_FIRST_HART and MPFS_HAL_LAST_HART in mss_sw_config file

Thanks for help
Kevin Wen

**Comment by hughbreslin** [comment, marked as answer]
Hey @kevinwen2000 

You can create four separate payloads as you described and set the MPFS_HAL_FIRST_HART and MPFS_HAL_LAST_HART values to be the same, for example to only run on U54_1:

MPFS_HAL_FIRST_HART = 1
MPFS_HAL_LAST_HART = 1

to only run on U54_2:

MPFS_HAL_FIRST_HART = 2
MPFS_HAL_LAST_HART = 2

There are a few yaml examples in the HSS, [this one](https://github.com/polarfire-soc/hart-software-services/blob/master/tools/hss-payload-generator/test/config.yaml) would probably be the most useful

**Comment by kevinwen2000** [comment]
Hello hughbreslin,

It works. I do what you said and 4 harts is running. But I think it is not easy to debug since it is coming from 4 independent projects.

I had tried to use single project for running 4 harts but failed. So far, I can run 3 harts with one project from ddr.

Thanks
Kevin
